### DISCUSSION TITLE
Can cvc5 be used directly for program synthesis by example?

### DISCUSSION THREAD

**Discussion by pauljurczak** [cvc5/cvc5#9791] [General]
I would like to synthesize a function `f` based on these examples:
```
f(3, 1) == 3
f(2, 2) == 2
f(1, 3) == 1
f(4, 4) == 4
f(8, 5) == 9
f(7, 6) == 9
f(9, 7) == 9
f(5, 8) == 9
f(6, 9) == 9
```
If needed, syntax constrains can be added. I'm expecting to get a short, simple function like this one:
```
def f(a, b):
  return a if b <= 4 else 9
```
Is this doable in cvc5? If not, what would be a tool to use, preferably a Python module?

**Comment by martin-cs** [comment]
Expression synthesis is part of cvc5's SyGuS solver functionality.  It should be able to handle this kind of thing.  If you don't know the SyGuS format then there are a bunch of examples in the regression tests:

https://github.com/cvc5/cvc5/tree/main/test/regress/cli/regress0/sygus

In particular

https://github.com/cvc5/cvc5/blob/main/test/regress/cli/regress0/sygus/issue4790-dtd.sy

looks like the kind of thing you might need.

HTH

**Reply by pauljurczak** [reply]
Thank you. I found an example (https://github.com/cvc5/cvc5/blob/main/examples/api/python/sygus-grammar.py), which substitutes SyGuS syntax with Python API? Since I haven't used SyGuS before, is there any downside in staying with this Python API substitute?

I think the API I mentioned is documented here: https://cvc5.github.io/docs/cvc5-1.0.2/api/python/base/grammar.html

I can't find Python API to read SyGuS format: https://stackoverflow.com/questions/76394657/how-to-read-sygus-format-into-cvc5-python-script

**Reply by martin-cs** [reply]
I haven't used the python API but my understanding is that it is pretty direct.  There will likely by a (small, linear) amount of overhead due to Python but there shouldn't be major drawbacks.  Or, at least, if you find one you should report it as a bug!

**Comment by pauljurczak** [comment]
I used this problem definition file `ex2.sy`:
```
(set-logic LIA)

(synth-fun f ((x Int) (y Int)) Int
    
    ((I Int) (B Bool))

    ((I Int (x y 0 1 2 3 4 5 6 7 8 9
             (+ I I) (- I I)
             (ite B I I)))
     (B Bool ((and B B) (or B B) (not B)
              (= I I) (<= I I) (>= I I))))
)

(declare-var x Int)
(declare-var y Int)

(constraint (= (f 3 1) 3))
(constraint (= (f 2 2) 2))
(constraint (= (f 1 3) 1))
(constraint (= (f 4 4) 4))
(constraint (= (f 8 5) 9))
(constraint (= (f 7 6) 9))
(constraint (= (f 9 7) 9))
(constraint (= (f 5 8) 9))
(constraint (= (f 6 9) 9))

(check-synth)
```
but the result of `./cvc5 ex2.sy` is somewhat convoluted:
```
(
(define-fun f ((x Int) (y Int)) Int (ite (= x y) x (ite (= x 1) x (ite (= x 3) x 9))))
)
```

**Reply by martin-cs** [reply]
Yes.  The goal is to synthesise /a/ function, not necessarily to synthesise the "best" or "simplest" function.  These are notoriously tricky things to formalise and even harder to implement.  There is possibly some scope to rewrite it after generation ( so maybe you get `(ite (or (= x y) (= x 1) (= x 3)) x 9)` ) but that's not going to get you to the `(ite (<= y 4) x 9)` you wanted initially.

**Reply by pauljurczak** [reply]
I asked specifically about it in https://github.com/cvc5/cvc5/discussions/9799. The suggestion to use `--no-sygus-pbe` produced the simplest function I was looking for.
